packages/scripts/src/upgrade-tests/Proposer.js

```diff
@@ -0,0 +1,137 @@
+
+const hre = require("hardhat");
+const Web3 = require("web3");
+
+// Create our own web3 to work around this hardhat bug: https://github.com/nomiclabs/hardhat/issues/1226.
+hre.web3 = new Web3(process.env.CUSTOM_NODE_URL);
+const { getContract, web3 } = hre;
+const { ZERO_ADDRESS } = require("@uma/common");
+const assert = require("assert");
+require("dotenv").config();
+const argv = require("minimist")(process.argv.slice(), {
+  string: [
+    // proposer address.
+    "proposer",
+  ],
+});
+
+const { run: simulateVote } = require("../admin-proposals/simulateVote");
+
+async function main() {
+  // Note: there are no default accounts, so all must be impersonated.
+  // This creates a random account as the default.
+  const defaultAccount = web3.utils.randomHex(20);
+  await addEth(defaultAccount);
+  await impersonateAccount(defaultAccount);
+
+  // Get contracts.
+  const Governor = getContract("Governor");
+  const Proposer = getContract("Proposer");
+  const VotingToken = getContract("VotingToken");
+  const Finder = getContract("Finder");
+  const ERC20 = getContract("ERC20");
+
+  const governor = await Governor.deployed();
+  const votingToken = await VotingToken.deployed();
+
+  // Deploy the proposer if not deployed.
+  let proposer;
+  if (argv.proposer) {
+    proposer = Proposer.at(argv.proposer);
+  } else {
+    const votingToken = await VotingToken.deployed();
+    const finder = await Finder.deployed();
+    proposer = await Proposer.new(
+      votingToken.options.address,
+      web3.utils.toWei("5000"),
+      governor.options.address,
+      finder.options.address,
+      ZERO_ADDRESS
+    ).send({ from: defaultAccount });
+  }
+
+  // Grab the current owner and proposer accounts from the governor.
+  const governorOwner = await governor.methods.getMember(0).call();
+  const governorProposer = await governor.methods.getMember(1).call();
+
+  // As with other accounts, the owner account will need to be given ETH and impersonated.
+  await addEth(governorOwner);
+  await impersonateAccount(governorOwner);
+
+  // Random large UMA holder (EOA). This account will be used as a source of UMA.
+  const largeHolder = "0x8bd16DE8938A2ad16794ac7E7502896f156E370C";
+  await impersonateAccount(largeHolder);
+
+  // Do upgrade to decentralized proposal system.
+  await governor.methods.resetMember(1, proposer.options.address).send({ from: governorOwner });
+  await governor.methods.resetMember(0, governor.options.address).send({ from: governorOwner });
+
+  // Send 10k tokens to the default account so it can make two proposals.
+  await votingToken.methods.transfer(defaultAccount, web3.utils.toWei("10000")).send({ from: largeHolder });
+
+  // Approve the proposal contract to take 10k tokens from the default wallet.
+  await votingToken.methods.approve(proposer.options.address, web3.utils.toWei("10000")).send({ from: defaultAccount });
+
+  // Create arbitrary transaction to test that the proposal system works for vanilla requests.
```

Comment:
this is good but I feel we should also have a more UMA specific test case. please add a proposal to add an identifier to the whitelist or approve a collateral or something else as well.

Reply:
Will add.

Reply:
Done.